{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "import os\n",
    "from pathlib import Path\n",
    "import re\n",
    "import pybedtools\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.stats import mannwhitneyu\n",
    "from collections import defaultdict\n",
    "import pycircos\n",
    "\n",
    "Garc = pycircos.Garc\n",
    "Gcircle = pycircos.Gcircle\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a dictionary to store DMR paths\n",
    "patient_dmrs = {}\n",
    "\n",
    "# Base directories\n",
    "input_directory = Path(\"path/to/input/directory\") # Change this to your directory\n",
    "results_directory = Path(\"path/to/output/directory\") # Change this to your directory\n",
    "\n",
    "# Subdirectories\n",
    "dmr_directory = input_directory/\"DMR_files\"\n",
    "cpg_directory = input_directory/\"CpG_files\"\n",
    "tss_directory = input_directory/\"TSS_files\"\n",
    "upwards_methylated_output_directory = results_directory/\"Upwards_methylated\"\n",
    "downwards_methylated_output_directory = results_directory/\"Downwards_methylated\"\n",
    "bed_output_directory = results_directory/\"BED_files\"\n",
    "cpg_and_tss_output_directory = results_directory/\"Corrected_CpGIs_\"\n",
    "\n",
    "# Columns to keep in the output DMR CSV files\n",
    "columns_to_keep = ['Chromosome', 'Start', 'End', 'SiteCount', 'Length', 'Percentage', 'FoldChange']\n",
    "\n",
    "# Create the output directories if they don't exist\n",
    "directories_to_create = [\n",
    "    results_directory,\n",
    "    upwards_methylated_output_directory,\n",
    "    downwards_methylated_output_directory,\n",
    "    bed_output_directory,\n",
    "    cpg_and_tss_output_directory\n",
    "]\n",
    "\n",
    "for directory in directories_to_create:\n",
    "    os.makedirs(directory, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the transformation function that corrects the original DMR files if the have any format flaws\n",
    "def transform_row(row):\n",
    "    # Check if Length is negative\n",
    "    if row['Length'] < 0:\n",
    "        # Make Length positive\n",
    "        row['Length'] = abs(row['Length'])\n",
    "        row['End'], row['Length'] = row['Length'], row['End']\n",
    "        row['Start'], row['End'] = row['End'], row['Start']\n",
    "    \n",
    "    return row\n",
    "\n",
    "# Function to short chromomosomes in the format 1,2,3...X,Y\n",
    "def chrom_sorter(chrom):\n",
    "    chrom = chrom[3:]  # Remove 'chr' prefix\n",
    "    if chrom.isdigit():  # For numeric chromosomes\n",
    "        return int(chrom)\n",
    "    elif chrom == 'X':\n",
    "        return 23\n",
    "    elif chrom == 'Y':\n",
    "        return 24\n",
    "    else:\n",
    "        return 25  # For any non-standard chromosome identifiers, sort them at the end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate over each DMR file in the input directory\n",
    "for filename in os.listdir(dmr_directory):\n",
    "    if filename.endswith('.csv'):\n",
    "        # Extract patient ID from the filename using regular expression\n",
    "        match = re.search(r'([NR|R]+)_(AZA|DAC)_(\\d+)', filename)\n",
    "        if match:\n",
    "            patient_id = match.group(1)\n",
    "            treatment = match.group(2)\n",
    "        else:\n",
    "            # Handle cases where the pattern doesn't match\n",
    "            print(f\"Could not extract patient ID and treatment from the filename: {filename}\")\n",
    "            continue\n",
    "        \n",
    "        # Specify input and output file paths\n",
    "        input_file_path = dmr_directory/filename\n",
    "        output_file_path = results_directory/filename\n",
    "\n",
    "        upwards_methylated_output_file_path = upwards_methylated_output_directory/f\"{os.path.splitext(filename)[0]}_upwards.csv\"\n",
    "        downwards_methylated_output_file_path = downwards_methylated_output_directory/f\"{os.path.splitext(filename)[0]}_downwards.csv\"\n",
    "\n",
    "        patient_dmrs[patient_id] = (output_file_path, upwards_methylated_output_file_path, downwards_methylated_output_file_path)\n",
    "\n",
    "        # Read the CSV file using pandas\n",
    "        df = pd.read_csv(input_file_path, delimiter=',', low_memory = False)\n",
    "\n",
    "        # Transform the rows\n",
    "        df = df.apply(transform_row, axis=1)\n",
    "\n",
    "        # Select specified columns and drop duplicates\n",
    "        df = df[columns_to_keep].drop_duplicates()\n",
    "        \n",
    "        # Convert percentage to float.\n",
    "        df['Percentage'] = df['Percentage'].astype(float)  # Replace commas with dots and convert to float\n",
    "\n",
    "        # Rename columns\n",
    "        df = df.rename(columns={\n",
    "            'Chromosome': 'chrom',\n",
    "            'Start': 'chromStart',\n",
    "            'End': 'chromEnd',\n",
    "            'SiteCount': 'LpnPI-sites',\n",
    "            'Percentage': 'methRatio',\n",
    "        })\n",
    "\n",
    "        # Log transform the methRatio column\n",
    "        df['log10(methRatio)'] = np.log10(df['methRatio'])\n",
    "\n",
    "        # Sort chromosomes in the order of chr1, chr2 ,chr3 and so on.\n",
    "        df['chromSort'] = df['chrom'].apply(chrom_sorter)\n",
    "        df = df.sort_values(by=['chromSort', 'chromStart']).drop('chromSort', axis=1)\n",
    "\n",
    "        # Apply filters to keep only regions with FoldChange >=2, meaning either an increase or decrease in methylation level by at least 100%.\n",
    "        # Also keep regions that have at least 20 LpnPI-sites, indicating significant methylation or demethylation.\n",
    "        df = df[(df['FoldChange'] >= 2) & (df['LpnPI-sites'] >= 20)]\n",
    "\n",
    "        # Remove chromosomes of alternative and random contig from the dataset.\n",
    "        df = df[df['chrom'].str.match(r'^chr((\\d{1,2})|(X|Y))$')]\n",
    "        df = df.drop(columns=['LpnPI-sites'])  # Drop the intermediate columns\n",
    "\n",
    "        # Split data into upwards and downwards methylated regions.\n",
    "        upwards_meth_df = df[df['log10(methRatio)'] > 0]\n",
    "        downwards_meth_df = df[df['log10(methRatio)'] < 0]\n",
    "\n",
    "        # Save files to csv.\n",
    "        upwards_meth_df.to_csv(upwards_methylated_output_file_path, index=False, sep=',')\n",
    "        downwards_meth_df.to_csv(downwards_methylated_output_file_path, index=False, sep=',')\n",
    "        df.to_csv(output_file_path, index=False, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create lists to group the CpGIs and TSS files for the 2 patients\n",
    "patient_6281 = [cpg_directory/\"6281_CpG.csv\", tss_directory/\"6281_TSS.csv\"]\n",
    "patient_5889 = [cpg_directory/\"5889_CpG.csv\", tss_directory/\"5889_TSS.csv\"]\n",
    "\n",
    "def format_file(input_file, output_directory):\n",
    "    # Read file into a pandas dataframe\n",
    "    df = pd.read_csv(input_file, low_memory=False)\n",
    "\n",
    "    # Rename columns early in the process\n",
    "    df = df.rename(columns={\n",
    "        'Chromosome': 'chrom',\n",
    "        'Start': 'chromStart',\n",
    "        'End': 'chromEnd',\n",
    "        'Percentage(Y/X)': 'methRatio',\n",
    "        'log10Score': 'log10(methRatio)'\n",
    "    })\n",
    "\n",
    "    # Convert the q-value column to numeric format and filter out rows with q-value higher than 0.05\n",
    "    df['q-value'] = pd.to_numeric(df['q-value'], errors='coerce')\n",
    "    df = df[df['q-value'] <= 0.05]\n",
    "\n",
    "    # Convert 'methRatio' and 'log10(methRatio)' columns to numeric, handling formatting issues\n",
    "    df['methRatio'] = pd.to_numeric(df['methRatio'].str.replace(',', ''), errors='coerce')\n",
    "    df['log10(methRatio)'] = pd.to_numeric(df['log10(methRatio)'], errors='coerce')\n",
    "\n",
    "    # Split the dataframe based on 'methRatio'\n",
    "    upwards_methylated = df[df['methRatio'] >= 1]\n",
    "    downward_methylated = df[df['methRatio'] < 1]\n",
    "\n",
    "    # Drop and recalibrate columns based on conditions\n",
    "    upwards_methylated = upwards_methylated.drop(columns=['methRatio'])\n",
    "    downward_methylated = downward_methylated.drop(columns=['log10(methRatio)'])\n",
    "    \n",
    "    upwards_methylated['methRatio'] = 10 ** upwards_methylated['log10(methRatio)']\n",
    "    downward_methylated['log10(methRatio)'] = np.log10(downward_methylated['methRatio'])\n",
    "\n",
    "    # Concatenate the adjusted dataframes\n",
    "    final_df = pd.concat([upwards_methylated, downward_methylated])\n",
    "    \n",
    "    # Remove chromosomes not matching the specified pattern\n",
    "    final_df = final_df[final_df['chrom'].str.match(r'^chr((\\d{1,2})|(X|Y))$')]\n",
    "\n",
    "    # Drop unused columns and rows with NaN in 'methRatio'\n",
    "    final_df = final_df.drop(columns=['Type', 'EnsembleID'])\n",
    "    final_df = final_df.dropna(subset=['methRatio'])\n",
    "\n",
    "    # Create the 'FoldChange' column\n",
    "    final_df['FoldChange'] = final_df['methRatio'].apply(lambda x: x if x >= 1 else 1/x)\n",
    "    final_df = final_df[final_df['FoldChange'] >= 2]\n",
    "\n",
    "    # Filter out rows based on the sum of ScoreX and ScoreY\n",
    "    final_df = final_df.loc[(final_df['ScoreX'] + final_df['ScoreY']) >= 20]\n",
    "\n",
    "    # Apply custom sorting function for chromosomes if available\n",
    "    final_df['chromSort'] = final_df['chrom'].apply(chrom_sorter)\n",
    "    final_df = final_df.sort_values(by=['chromSort', 'chromStart']).drop('chromSort', axis=1)\n",
    "\n",
    "    # Specify and apply a new column order\n",
    "    new_column_order = ['chrom', 'chromStart', 'chromEnd', 'ScoreX', 'ScoreY', 'methRatio',\n",
    "                         'FoldChange', 'log10(methRatio)', 'FDR', 'q-value']\n",
    "    final_df = final_df[new_column_order]\n",
    "\n",
    "    # Construct output filename and save the processed DataFrame\n",
    "    base_filename = os.path.basename(input_file)\n",
    "    output_filename = f\"corrected_{base_filename}\"\n",
    "    output_file_path = output_directory/output_filename\n",
    "    final_df.to_csv(output_file_path, index=False)\n",
    "    final_df = final_df.drop(columns=['FDR','q-value'])\n",
    "\n",
    "    return final_df\n",
    "\n",
    "# Format each file in the lists and concatanate the 2 files for each patient to a final dataframe.\n",
    "df_6281_list = [format_file(file_path, cpg_and_tss_output_directory) for file_path in patient_6281]\n",
    "final_df_6281 = pd.concat(df_6281_list)\n",
    "\n",
    "df_5889_list = [format_file(file_path, cpg_and_tss_output_directory) for file_path in patient_5889]\n",
    "final_df_5889 = pd.concat(df_5889_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to detect Differentially Methylated Regions within the CpGIs and TSS sites.\n",
    "def find_DMRs(df, distance_threshold, methylation_threshold):\n",
    "    # Sort the input DataFrame by chromosome ('chrom') and the start position ('chromStart') of the genetic regions.\n",
    "    df_sorted = df.sort_values(by=['chrom', 'chromStart'])\n",
    "    DMRs = []  # Initialize an empty list to store identified DMRs.\n",
    "\n",
    "    # Group the sorted DataFrame by chromosome and iterate over each group.\n",
    "    for chrom, group in df_sorted.groupby('chrom'):\n",
    "        # Initialize the first DMR with the first row's data\n",
    "        current_DMR = {\n",
    "            'chrom': chrom,\n",
    "            'start': group.iloc[0]['chromStart'],\n",
    "            'end': group.iloc[0]['chromEnd'],\n",
    "            'total_log10MethRatio': group.iloc[0]['log10(methRatio)'],\n",
    "            'total_FoldChange': group.iloc[0]['FoldChange'],\n",
    "            'total_MethRatio': group.iloc[0]['methRatio'],\n",
    "            'count': 1\n",
    "        }\n",
    "\n",
    "        # Iterate over each row in the group to assess and aggregate contiguous regions into DMRs.\n",
    "        for _, row in group.iterrows():\n",
    "            within_distance = row['chromStart'] <= current_DMR['start'] + distance_threshold\n",
    "            avg_log10MethRatio = current_DMR['total_log10MethRatio'] / current_DMR['count']\n",
    "            log10methRatio_diff = abs(avg_log10MethRatio - row['log10(methRatio)'])\n",
    "\n",
    "            if within_distance and log10methRatio_diff <= methylation_threshold:\n",
    "                # Aggregate the region into the current DMR if within distance and threshold.\n",
    "                current_DMR['start'] = min(current_DMR['start'], row['chromStart'])\n",
    "                current_DMR['end'] = max(current_DMR['end'], row['chromEnd'])\n",
    "                current_DMR['total_log10MethRatio'] += row['log10(methRatio)']\n",
    "                current_DMR['total_FoldChange'] += row['FoldChange']\n",
    "                current_DMR['total_MethRatio'] += row['methRatio']\n",
    "                current_DMR['count'] += 1\n",
    "            else:\n",
    "                # Finalize the current DMR and start a new one.\n",
    "                DMRs.append([\n",
    "                    current_DMR['chrom'], current_DMR['start'], current_DMR['end'], current_DMR['total_log10MethRatio'],\n",
    "                    current_DMR['total_FoldChange'], current_DMR['total_MethRatio'], current_DMR['count']\n",
    "                ])\n",
    "                current_DMR = {\n",
    "                    'chrom': chrom,\n",
    "                    'start': row['chromStart'],\n",
    "                    'end': row['chromEnd'],\n",
    "                    'total_log10MethRatio': row['log10(methRatio)'],\n",
    "                    'total_FoldChange': row['FoldChange'],\n",
    "                    'total_MethRatio': row['methRatio'],\n",
    "                    'count': 1\n",
    "                }\n",
    "\n",
    "        # Append the last DMR for the chromosome\n",
    "        DMRs.append([\n",
    "            current_DMR['chrom'], current_DMR['start'], current_DMR['end'], current_DMR['total_log10MethRatio'],\n",
    "            current_DMR['total_FoldChange'], current_DMR['total_MethRatio'], current_DMR['count']\n",
    "        ])\n",
    "\n",
    "    # Convert the list of DMRs to a DataFrame and calculate additional metrics.\n",
    "    DMRs_df = pd.DataFrame(DMRs, columns=['Chrom', 'Start', 'End', 'Total_log10MethRatio', 'Total_FoldChange', 'Total_MethRatio', 'Count'])\n",
    "    DMRs_df['Length'] = DMRs_df['End'] - DMRs_df['Start']\n",
    "    DMRs_df['Average_MethRatio'] = DMRs_df['Total_MethRatio'] / DMRs_df['Count']\n",
    "    DMRs_df['Average_FoldChange'] = DMRs_df['Total_FoldChange'] / DMRs_df['Count']\n",
    "    DMRs_df['Average_log10MethRatio'] = DMRs_df['Total_log10MethRatio'] / DMRs_df['Count']\n",
    "\n",
    "    # Reorder and rename columns to match the requested format.\n",
    "    DMRs_df = DMRs_df[['Chrom', 'Start', 'End', 'Length', 'Average_MethRatio', 'Average_FoldChange', 'Average_log10MethRatio']]\n",
    "    DMRs_df.columns = ['chrom', 'chromStart', 'chromEnd', 'Length', 'methRatio', 'FoldChange', 'log10(methRatio)']\n",
    "    DMRs_df = DMRs_df.drop(columns=['log10(methRatio)'])\n",
    "\n",
    "    DMRs_df['log10(methRatio)'] = np.log10(DMRs_df['methRatio'])\n",
    "\n",
    "    # Convert relevant columns to integer type.\n",
    "    DMRs_df[['chromStart', 'chromEnd', 'Length']] = DMRs_df[['chromStart', 'chromEnd', 'Length']].astype(int)\n",
    "\n",
    "    return DMRs_df\n",
    "\n",
    "# Define a function to create a final file for each patient.\n",
    "def DMR_file_creation(df, upwards_meth_filename, downwards_meth_filename, final_DMR_filename):\n",
    "    # Create upwards and downwards methylated dataframes based on log10(methRatio)\n",
    "    upwards_meth_df = df[df['log10(methRatio)'] >= 0]\n",
    "    downwards_meth_df = df[df['log10(methRatio)'] < 0]\n",
    "\n",
    "    # Apply the function for the DMR detection to all files\n",
    "    upwards_meth_df = find_DMRs(upwards_meth_df, distance_threshold = 50000, methylation_threshold = 10)\n",
    "    upwards_meth_df.to_csv(upwards_meth_filename, index = False)\n",
    " \n",
    "    downwards_meth_df = find_DMRs(downwards_meth_df, distance_threshold = 50000, methylation_threshold = 10)\n",
    "    downwards_meth_df.to_csv(downwards_meth_filename, index = False)\n",
    "    \n",
    "    # Concatanate and sort DMRs by chromosome.\n",
    "    final_DMR_file = pd.concat([upwards_meth_df, downwards_meth_df])\n",
    "    final_DMR_file = final_DMR_file.sort_values(by=['chrom', 'chromStart'], ignore_index= True)\n",
    "\n",
    "    # Fit and transform the winsorized data with RobustScaler\n",
    "    final_DMR_file.to_csv(final_DMR_filename, index = False)\n",
    "\n",
    "    # Return the final DMR file\n",
    "    return final_DMR_file\n",
    "\n",
    "# Create the DMR file for both patients.\n",
    "final_DMR_file_5889 = DMR_file_creation(final_df_5889, upwards_methylated_output_directory/\"R_AZA_5889_upwards.csv\", downwards_methylated_output_directory/\"R_AZA_5889_downwards.csv\", results_directory/\"R_AZA_5889.csv\")\n",
    "final_DMR_file_6281 = DMR_file_creation(final_df_6281, upwards_methylated_output_directory/\"R_AZA_6281_upwards.csv\", downwards_methylated_output_directory/\"R_AZA_6281_downwards.csv\", results_directory/\"R_AZA_6281.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the lists to store dataframes and their identifiers\n",
    "AZA_non_responders = []\n",
    "AZA_responders = []\n",
    "DAC_non_responders = []\n",
    "\n",
    "# List all files in the data directory\n",
    "for filename in os.listdir(results_directory):\n",
    "    if filename.endswith('.csv'):\n",
    "        # Extract the identifier\n",
    "        identifier = filename.split('_')[-1].split('.')[0]\n",
    "        \n",
    "        # Load the dataset\n",
    "        df = pd.read_csv(os.path.join(results_directory, filename))\n",
    "        \n",
    "        # Determine the category of the dataset and append it to the corresponding list\n",
    "        if 'NR_AZA' in filename:\n",
    "            AZA_non_responders.append((identifier, df))\n",
    "        elif 'R_AZA' in filename:\n",
    "            AZA_responders.append((identifier, df))\n",
    "        elif 'NR_DAC' in filename:\n",
    "            DAC_non_responders.append((identifier, df))\n",
    "\n",
    "plots_directory = results_directory / \"Plots\"\n",
    "\n",
    "# Check if results_directory exists, if not, create it\n",
    "if not os.path.exists(plots_directory):\n",
    "    os.makedirs(plots_directory)\n",
    "\n",
    "# Define a function to plot different metrics of the methRatio scores\n",
    "def plot_combined_metrics(df, category, plots_directory):\n",
    "    # Set up the matplotlib figure\n",
    "    fig, axs = plt.subplots(2, 2, figsize=(16, 12))  # Adjust size as needed\n",
    "    \n",
    "    # First plot - methRatio\n",
    "    sns.histplot(df['methRatio'], bins=100, color='skyblue', kde=True, ax=axs[0, 0])\n",
    "    axs[0, 0].set_title(f'Methylation ratio (methRatio) for {category}_{identifier}')\n",
    "    axs[0, 0].set_xlabel('Methylation Ratio')\n",
    "    axs[0, 0].set_ylabel('Frequency')\n",
    "    \n",
    "    # Second plot - log10(methRatio)\n",
    "    sns.histplot(df['log10(methRatio)'], bins=100, color='orange', kde=True, ax=axs[0, 1])\n",
    "    axs[0, 1].set_title(f'Log10(methRatio) for {category}_{identifier}')\n",
    "    axs[0, 1].set_xlabel('Log10(methRatio)')\n",
    "    axs[0, 1].set_ylabel('Frequency')\n",
    "    \n",
    "    # Adjust layout\n",
    "    plt.tight_layout()\n",
    "    sns.despine\n",
    "\n",
    "    # Save the plot to file\n",
    "    plot_output_path = plots_directory / f\"{category}_{identifier}_combined_metrics_plot.png\"\n",
    "    plt.savefig(plot_output_path)\n",
    "    plt.close(fig)  # Close the figure to free up memory\n",
    "\n",
    "# Define a dictionary 'categories' with keys representing different patient categories and values being lists of tuples. \n",
    "# Each tuple contains an identifier for a patient or sample and a DataFrame ('df') with their data.\n",
    "response_categories = {\n",
    "    'NR_AZA': AZA_non_responders,    # Non-responders to Azacitidine treatment\n",
    "    'R_AZA': AZA_responders,         # Responders to Azacitidine treatment\n",
    "    'NR_DAC': DAC_non_responders    # Non-responders to Decitabine treatment\n",
    "}\n",
    "\n",
    "# Iterate over each category and its corresponding list of data tuples in the 'categories' dictionary.\n",
    "for category, data_list in response_categories.items():\n",
    "    # Iterate over each tuple in the data list for the current category.\n",
    "    # 'identifier' is a unique identifier for the patient or sample,\n",
    "    # and 'df' is the DataFrame containing their methylation data.\n",
    "    for identifier, df in data_list:\n",
    "        # Construct a filename for saving the plot, using the unique 'identifier'.\n",
    "        # This avoids category duplication in the filename, making it more concise.\n",
    "        filename = f'{identifier}'\n",
    "        \n",
    "        # Call the function 'plot_log10methratio' to generate and save a plot of the log10-transformed methylation ratio data contained in 'df'. \n",
    "        # The plot is saved with the constructed 'filename', and 'category' is passed to potentially customize the plot based on the patient category.\n",
    "        plot_combined_metrics(df, category, plots_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the weighted mean log10-transformed methylation ratio for each chromosome within a DataFrame.\n",
    "def calculate_weighted_mean_methRatio(df, identifier):\n",
    "    # Initialize a DataFrame to store the weighted means and weighted standard errors.\n",
    "    results_list = []\n",
    "\n",
    "    # Iterate through each chromosome in the dataframe\n",
    "    for chrom in df['chrom'].unique():\n",
    "        # Extract the subset of the dataframe for the chromosome.\n",
    "        chrom_df = df[df['chrom'] == chrom]\n",
    "        \n",
    "        # Split the data into two groups: above and below zero.\n",
    "        above_zero = chrom_df[chrom_df['log10(methRatio)'] > 0]\n",
    "        below_zero = chrom_df[chrom_df['log10(methRatio)'] < 0]\n",
    "        \n",
    "        # Calculate the mean of each group.\n",
    "        mean_above_zero = above_zero['log10(methRatio)'].mean() if not above_zero.empty else 0\n",
    "        mean_below_zero = below_zero['log10(methRatio)'].mean() if not below_zero.empty else 0\n",
    "        \n",
    "        # Calculate the weight (percentage) of each group.\n",
    "        weight_above_zero = len(above_zero) / len(chrom_df) if not chrom_df.empty else 0\n",
    "        weight_below_zero = len(below_zero) / len(chrom_df) if not chrom_df.empty else 0\n",
    "        \n",
    "        # Calculate the standard error for each group.\n",
    "        sem_above_zero = above_zero['log10(methRatio)'].std() / np.sqrt(len(above_zero)) if not above_zero.empty else 0\n",
    "        sem_below_zero = below_zero['log10(methRatio)'].std() / np.sqrt(len(below_zero)) if not below_zero.empty else 0\n",
    "        \n",
    "        # Calculate the weighted mean for the chromosome.\n",
    "        weighted_mean = (mean_above_zero * weight_above_zero) + (mean_below_zero * weight_below_zero)\n",
    "        \n",
    "        # Calculate the weighted standard error.\n",
    "        # Here, we assume independence between groups for simplicity.\n",
    "        weighted_sem = np.sqrt((sem_above_zero ** 2 * weight_above_zero) + (sem_below_zero ** 2 * weight_below_zero))\n",
    "        \n",
    "        # Append the results to the list.\n",
    "        results_list.append({\n",
    "            'chrom': chrom,\n",
    "            'weighted_mean_log10_methRatio': weighted_mean,\n",
    "            'weighted_sem_log10_methRatio': weighted_sem,\n",
    "            'Identifier': identifier\n",
    "        })\n",
    "    \n",
    "    # Convert the list of dictionaries to a DataFrame.\n",
    "    df_results = pd.DataFrame(results_list)\n",
    "    \n",
    "    return df_results\n",
    "\n",
    "# Generate and save a bar plot visualizing mean methylation by chromosome for each patient ID within a category.\n",
    "def save_patient_barplot(data, category, chrom_order, plots_directory):\n",
    "    plt.figure(figsize=(14, 8))\n",
    "    ax = sns.barplot(\n",
    "        data=data,\n",
    "        x='chrom',\n",
    "        y='weighted_mean_log10_methRatio',\n",
    "        hue='Identifier',  # Differentiate data by patient ID.\n",
    "        errorbar='sd',  # Include standard deviation as error bars.\n",
    "        palette='deep',  # Use a deep color palette for distinction.\n",
    "        width=0.9,  # Bar width\n",
    "        edgecolor='black',  # Set the color of the bar border.\n",
    "        linewidth=0.7  # Set the width of the bar border.\n",
    "    )\n",
    "    \n",
    "    # Apply alternating background shading for chromosomes to enhance readability.\n",
    "    for i, chrom in enumerate(chrom_order):\n",
    "        if i % 2 == 0:\n",
    "            plt.axvspan(i - 0.5, i + 0.5, color='grey', alpha=0.2)\n",
    "    \n",
    "    # Set plot title, labels, and other aesthetic elements.\n",
    "    plt.title(f'Median methylation by Chromosome for {category}', fontsize=16)\n",
    "    ax.set_xlabel('Chromosome', fontsize=14)\n",
    "    ax.set_ylabel('Median methylation', fontsize=14)\n",
    "    plt.xticks(rotation=35)  # Rotate chromosome labels for clarity.\n",
    "    plt.grid(True, axis='y')  # Enable grid lines along the y-axis.\n",
    "    plt.legend(title='Patient ID', bbox_to_anchor=(1, 1), loc='upper left', fontsize=12)\n",
    "    sns.despine()  # Remove top and right borders for a cleaner look.\n",
    "    plt.tight_layout()  # Adjust layout to ensure everything fits without overlap.\n",
    "    \n",
    "    # Save the plot to a file, naming it uniquely based on the category and patient IDs.\n",
    "    plot_output_path = plots_directory/f\"{category}_weighted_mean_chrom_barplot.png\"\n",
    "    plt.savefig(plot_output_path)\n",
    "    plt.close()\n",
    "\n",
    "# Prepare a list and categorical data type for chromosome ordering.\n",
    "chromosomes = ['chr' + str(i) for i in range(1, 23)] + ['chrX', 'chrY']\n",
    "chrom_order = pd.Categorical(chromosomes, categories=chromosomes, ordered=True)\n",
    "\n",
    "all_data = []  # Initialize a list to collect data for plotting.\n",
    "\n",
    "# Iterate over categories and their respective data lists to calculate mean methylation ratios.\n",
    "for category, data_list in response_categories.items():\n",
    "    for identifier, df in data_list:\n",
    "        df_weighted_mean = calculate_weighted_mean_methRatio(df, identifier)\n",
    "        df_weighted_mean['Category'] = category  # Label the data with its category.\n",
    "        all_data.append(df_weighted_mean)\n",
    "\n",
    "# Concatenate all collected data into a single DataFrame for plotting.\n",
    "plot_data_df = pd.concat(all_data)\n",
    "\n",
    "# Ensure the 'chrom' column is treated as categorical with a defined order, facilitating correct plot ordering.\n",
    "plot_data_df['chrom'] = plot_data_df['chrom'].astype('category')\n",
    "plot_data_df['chrom'] = plot_data_df['chrom'].cat.set_categories(chromosomes, ordered=True)\n",
    "\n",
    "# Generate and save bar plots for each category, visualizing mean methylation by chromosome.\n",
    "for category in plot_data_df['Category'].unique():\n",
    "    category_data = plot_data_df[plot_data_df['Category'] == category]\n",
    "    save_patient_barplot(category_data, category, chromosomes, plots_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to create a DataFrame that contains all samples for each category and adds a category column.\n",
    "def concat_dataframes_from_tuples(tuples_list, category_name):\n",
    "    # Extract just the DataFrame from each tuple and add a 'category' column\n",
    "    dfs = []\n",
    "    for _, df in tuples_list:\n",
    "        # Make a copy to avoid changing the original DataFrame\n",
    "        df_copy = df.copy()\n",
    "        df_copy['Category'] = category_name  # Add a new column with the category name\n",
    "        dfs.append(df_copy)\n",
    "        \n",
    "    # Concatenate all DataFrames in the list into one, if not empty\n",
    "    if dfs:\n",
    "        return pd.concat(dfs, ignore_index=True)\n",
    "    else:\n",
    "        return pd.DataFrame()  # Return an empty DataFrame if the list is empty\n",
    "\n",
    "# Combine NR_AZA and NR_DAC into one list and rename categories\n",
    "combined_non_responders = AZA_non_responders + DAC_non_responders\n",
    "\n",
    "# Use the modified function to concatenate DataFrames and add the 'category' column\n",
    "responders_df = concat_dataframes_from_tuples(AZA_responders, 'Responders')\n",
    "non_responders_df = concat_dataframes_from_tuples(combined_non_responders, 'Non_Responders')\n",
    "\n",
    "final_df = pd.concat([responders_df, non_responders_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the weighted means for the responders and non responders.\n",
    "responder_weighted_means = calculate_weighted_mean_methRatio(responders_df, identifier=\"Responders\")\n",
    "non_responder_weighted_means = calculate_weighted_mean_methRatio(non_responders_df, identifier=\"Non_Responders\")\n",
    "\n",
    "# Combine the weighted means in on dataframe for plotting.\n",
    "combined_weighted_means = pd.concat([responder_weighted_means, non_responder_weighted_means], ignore_index=True)\n",
    "combined_weighted_means['chrom'] = pd.Categorical(combined_weighted_means['chrom'], categories=chrom_order, ordered=True)\n",
    "\n",
    "plt.figure(figsize=(16, 8))\n",
    "\n",
    "# Define bar width and positions\n",
    "bar_width = 0.45\n",
    "index = np.arange(len(chrom_order))\n",
    "\n",
    "# Create a color map for your identifiers\n",
    "color_map = {'Responders': '#008000', 'Non_Responders': '#d62728'}\n",
    "\n",
    "# Assuming combined_weighted_means is your DataFrame\n",
    "# Example DataFrame setup remains the same...\n",
    "\n",
    "plt.figure(figsize=(16, 8))\n",
    "\n",
    "# Define the offset from the center for each bar\n",
    "# This moves the first bar to the left of the tick and the second bar to the right\n",
    "offset = bar_width / 2\n",
    "\n",
    "# Plot each group with adjusted positions\n",
    "for i, identifier in enumerate(['Responders', 'Non_Responders']):\n",
    "    group_df = combined_weighted_means[combined_weighted_means['Identifier'] == identifier]\n",
    "    means = group_df.groupby('chrom', observed=True)['weighted_mean_log10_methRatio'].mean()\n",
    "    sems = group_df.groupby('chrom', observed=True)['weighted_sem_log10_methRatio'].mean()\n",
    "\n",
    "    # Adjust the position based on the identifier index (i)\n",
    "    # This moves the bars so that one is on the left and one is on the right of the tick mark\n",
    "    if i == 0:  # Responders\n",
    "        pos = index - offset\n",
    "    else:  # Non_Responders\n",
    "        pos = index + offset\n",
    "\n",
    "    plt.bar(pos, means.reindex(chrom_order), bar_width, yerr=sems.reindex(chrom_order), color=color_map[identifier], label=identifier, capsize=5, edgecolor='black', linewidth=0.7)\n",
    "\n",
    "# Alternating background shading\n",
    "for i, chrom in enumerate(chrom_order):\n",
    "    if i % 2 == 0:\n",
    "        plt.axvspan(i - 0.5, i + 0.5, color='grey', alpha=0.2)\n",
    "\n",
    "plt.title(' ', fontsize=16)\n",
    "plt.xlabel(' ', fontsize=14)\n",
    "plt.ylabel('Weighted mean methylation', fontsize=16)\n",
    "plt.xticks(range(len(chrom_order)), chrom_order, fontsize=14, rotation=35)\n",
    "plt.yticks(fontsize=14)\n",
    "plt.grid(True, which='both', linestyle='--', linewidth=0.5)\n",
    "plt.legend(bbox_to_anchor=(0.93, 1), loc='upper left', fontsize=12)\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plot_output_path = plots_directory/\"weighted_mean_chrom_barplot.jpeg\"\n",
    "plt.savefig(plot_output_path, dpi=300)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy the final dataframe in order to make modifications only the data for the violin plot\n",
    "violin_plot_data = final_df\n",
    "violin_plot_data['chrom'] = pd.Categorical(violin_plot_data['chrom'], categories=chrom_order, ordered=True)\n",
    "\n",
    "# Set the size of the plot\n",
    "plt.figure(figsize=(16, 8))\n",
    "\n",
    "# Create violin plots\n",
    "sns.violinplot(x='chrom', y='log10(methRatio)', hue='Category', data=violin_plot_data, split=True, inner=\"quart\", density_norm='width', palette={'Responders': '#008000', 'Non_Responders': 'red'})\n",
    "# Alternating background shading\n",
    "for i, chrom in enumerate(chrom_order):\n",
    "    if i % 2 == 0:\n",
    "        plt.axvspan(i - 0.5, i + 0.5, color='grey', alpha=0.2)\n",
    "# Improve plot aesthetics and readability\n",
    "plt.title(' ',fontsize=16)\n",
    "plt.xticks(fontsize=14, rotation=35)\n",
    "plt.yticks(fontsize=14)\n",
    "plt.xlabel(' ', fontsize=14)\n",
    "plt.ylabel('Distribution of methylation', fontsize=16)\n",
    "plt.grid(True, which='both', linestyle='--', linewidth=0.5)\n",
    "plt.legend(bbox_to_anchor=(0.933, 1), loc='upper left', fontsize=12)\n",
    "sns.despine() \n",
    "plot_output_path = plots_directory/\"violin_plot.jpeg\"\n",
    "plt.tight_layout()\n",
    "plt.savefig(plot_output_path, dpi=300)\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the function to process and split dataframes\n",
    "def process_and_split_df(df):\n",
    "    # Drop specified columns\n",
    "    df = df.drop(columns=['Length', 'methRatio', 'FoldChange'])\n",
    "    \n",
    "    # Split the dataframe based on the log10methratio threshold\n",
    "    upmeth_df = df[df['log10(methRatio)'] >= 0 ]\n",
    "    downmeth_df = df[df['log10(methRatio)'] < 0 ]\n",
    "    \n",
    "    return upmeth_df, downmeth_df\n",
    "\n",
    "# Define the function to convert a dataframe to a BED file\n",
    "def convert_df_to_bed(df, output_file_name):\n",
    "    # Assuming the dataframe has the necessary columns for BED format in the correct order\n",
    "    bed = pybedtools.BedTool.from_dataframe(df)\n",
    "    bed.saveas(output_file_name)\n",
    "\n",
    "# Process and convert dataframes in each list\n",
    "def create_BED_files(dataframe_lists, output_directory):\n",
    "    for list_name, dataframes in dataframe_lists.items():\n",
    "        for identifier, df in dataframes:\n",
    "            # Process and split the dataframe\n",
    "            upmeth_df, downmeth_df = process_and_split_df(df)\n",
    "            \n",
    "            # Convert each split dataframe to a BED file\n",
    "            upmeth_bed_file = os.path.join(output_directory, f\"upwards_{list_name}_{identifier}.bed\")\n",
    "            convert_df_to_bed(upmeth_df, upmeth_bed_file)\n",
    "            \n",
    "            downmeth_bed_file = os.path.join(output_directory, f\"downwards_{list_name}_{identifier}.bed\")\n",
    "            convert_df_to_bed(downmeth_df, downmeth_bed_file)\n",
    "\n",
    "\n",
    "# Process all dataframes\n",
    "create_BED_files(response_categories, bed_output_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# At this point we will split each chromosome into bins of the same specific length. \n",
    "# Exact lengths for each human chromosome based on hg38.\n",
    "chromosome_lengths = {\n",
    "    'chr1': 248956422, 'chr2': 242193529, 'chr3': 198295559,\n",
    "    'chr4': 190214555, 'chr5': 181538259, 'chr6': 170805979,\n",
    "    'chr7': 159345973, 'chr8': 145138636, 'chr9': 138394717,\n",
    "    'chr10': 133797422, 'chr11': 135086622, 'chr12': 133275309,\n",
    "    'chr13': 114364328, 'chr14': 107043718, 'chr15': 101991189,\n",
    "    'chr16': 90338345, 'chr17': 83257441, 'chr18': 80373285,\n",
    "    'chr19': 58617616, 'chr20': 64444167, 'chr21': 46709983,\n",
    "    'chr22': 50818468, 'chrX': 156040895, 'chrY': 57227415\n",
    "}\n",
    "\n",
    "# Define bin size\n",
    "bin_size = 100000  # 100kb \n",
    "\n",
    "# Initialize an empty list to hold the bin information\n",
    "bins_list = []\n",
    "\n",
    "# Generate bins for each chromosome\n",
    "for chrom, length in chromosome_lengths.items():\n",
    "    num_bins = length // bin_size + (1 if length % bin_size > 0 else 0)\n",
    "    for bin_num in range(num_bins):\n",
    "        bin_start = bin_num * bin_size\n",
    "        bin_end = min((bin_num + 1) * bin_size - 1, length)  # Ensure the last bin doesn't exceed chromosome length\n",
    "        bin_name = f\"{chrom}_bin{bin_num + 1}\"  # Assign bin names indicating the chromosome and bin number\n",
    "        bins_list.append({'chrom': chrom, 'bin_start': bin_start, 'bin_end': bin_end, 'bin_name': bin_name})\n",
    "\n",
    "# Convert the list of bins to a DataFrame\n",
    "bins_df = pd.DataFrame(bins_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating an IntervalIndex for genomic bins for each chromosome for efficiency\n",
    "# This dictionary comprehension creates an IntervalIndex for each chromosome found in bins_df.\n",
    "# It allows for efficient interval queries, such as finding overlaps with genomic regions.\n",
    "bins_intervals = {\n",
    "    chrom: pd.IntervalIndex.from_arrays(\n",
    "        bins_df[bins_df['chrom'] == chrom]['bin_start'],  # Start positions of bins for the chromosome\n",
    "        bins_df[bins_df['chrom'] == chrom]['bin_end'],    # End positions of bins for the chromosome\n",
    "        closed='both'  # Indicates that the intervals include both start and end points\n",
    "    )\n",
    "    for chrom in bins_df['chrom'].unique()  # Iterates over each unique chromosome in the dataframe\n",
    "}\n",
    "\n",
    "# Precompute a mapping from (chromosome, start, end) to bin name for efficient lookup\n",
    "# This dictionary maps a tuple of chromosome, start, and end positions to the corresponding bin name.\n",
    "# It's useful for quickly finding the name of the bin that a genomic region belongs to.\n",
    "bin_name_map = {\n",
    "    (row['chrom'], row['bin_start'], row['bin_end']): row['bin_name']  # Key: Tuple of chrom, start, end; Value: bin name\n",
    "    for _, row in bins_df.iterrows()  # Iterates over each row in the dataframe\n",
    "}\n",
    "\n",
    "# Define a function to assign bins with maximum overlap to genomic regions in a dataframe\n",
    "def assign_bins_with_max_overlap(df, bins_intervals, bin_name_map):\n",
    "    # Check if the input dataframe contains the necessary columns\n",
    "    if not {'chrom', 'chromStart', 'chromEnd'}.issubset(df.columns):\n",
    "        raise ValueError(\"Input DataFrame must contain 'chrom', 'chromStart', and 'chromEnd' columns.\")\n",
    "    \n",
    "    df['bin'] = 'No match'  # Initialize a new column for the assigned bins with a default 'No match' value\n",
    "\n",
    "    # Process each chromosome present in the input dataframe\n",
    "    for chrom in df['chrom'].unique():\n",
    "        if chrom not in bins_intervals:  # Skip chromosomes not in the bins_intervals dictionary\n",
    "            continue\n",
    "\n",
    "        # Filter the input dataframe for the current chromosome\n",
    "        chrom_df = df[df['chrom'] == chrom]\n",
    "        # Retrieve the IntervalIndex for the current chromosome\n",
    "        intervals = bins_intervals[chrom]\n",
    "\n",
    "        # Iterate over each genomic region in chrom_df\n",
    "        for index, row in chrom_df.iterrows():\n",
    "            # Create an interval for the current genomic region\n",
    "            region_interval = pd.Interval(left=row['chromStart'], right=row['chromEnd'], closed='both')\n",
    "            # Find bins that overlap with the region_interval\n",
    "            overlapping_bins = intervals.overlaps(region_interval)\n",
    "\n",
    "            # Skip if no bins overlap with the region\n",
    "            if not any(overlapping_bins):\n",
    "                continue\n",
    "\n",
    "            # Calculate overlap lengths for each overlapping bin and select the bin with maximum overlap\n",
    "            overlap_lengths = [\n",
    "                (min(row['chromEnd'], bin_index.right) - max(row['chromStart'], bin_index.left), bin_index.left, bin_index.right)\n",
    "                for bin_index in intervals[overlapping_bins]  # Iterate over intervals that overlap with the region\n",
    "            ]\n",
    "            # Select the interval with the maximum overlap\n",
    "            _, max_overlap_start, max_overlap_end = max(overlap_lengths, key=lambda x: x[0])\n",
    "            # Lookup the bin name using the max overlapping interval's start and end\n",
    "            bin = bin_name_map.get((chrom, max_overlap_start, max_overlap_end), 'No match')\n",
    "            # Update the 'bin' column in the original dataframe\n",
    "            df.at[index, 'bin'] = bin\n",
    "\n",
    "    return df  # Return the modified dataframe with bins assigned\n",
    "\n",
    "# Apply the bin detection function to each dataframe.\n",
    "for category, data_tuples in response_categories.items():\n",
    "    processed_dataframes = []\n",
    "    for identifier, df in data_tuples:\n",
    "        processed_df = assign_bins_with_max_overlap(df, bins_intervals, bin_name_map)\n",
    "        processed_dataframes.append((identifier, processed_df))\n",
    "    response_categories[category] = processed_dataframes\n",
    "\n",
    "methylation_df = assign_bins_with_max_overlap(final_df, bins_intervals, bin_name_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to perform Mann Whitney test.\n",
    "def perform_mannwhitney_u_test(df, significance_levels={'***': 0.001, '**': 0.01, '*': 0.05}):\n",
    "    # Identify bins with both categories present\n",
    "    bins_with_both_categories = df.groupby('bin')['Category'].nunique()\n",
    "    bins_with_both = bins_with_both_categories[bins_with_both_categories == 2].index.tolist()\n",
    "    \n",
    "    # Filter original dataset for these bins\n",
    "    df_filtered = df[df['bin'].isin(bins_with_both)]\n",
    "    \n",
    "    # Initialize results storage\n",
    "    test_results = []\n",
    "    \n",
    "    # Perform Mann-Whitney U tests and determine significance\n",
    "    for bin_name in bins_with_both:\n",
    "        data_resp = df_filtered[(df_filtered['Category'] == 'Responders') & (df_filtered['bin'] == bin_name)]['log10(methRatio)']\n",
    "        data_non_resp = df_filtered[(df_filtered['Category'] == 'Non_Responders') & (df_filtered['bin'] == bin_name)]['log10(methRatio)']\n",
    "        \n",
    "        # Perform Mann-Whitney U test\n",
    "        stat, p_value = mannwhitneyu(data_resp, data_non_resp, alternative='two-sided')\n",
    "        \n",
    "        # Determine significance annotation\n",
    "        annotation = ''\n",
    "        for symbol, threshold in significance_levels.items():\n",
    "            if p_value < threshold:\n",
    "                annotation = symbol\n",
    "                break\n",
    "        \n",
    "        test_results.append((bin_name, p_value, annotation))\n",
    "    \n",
    "    # Convert results to DataFrame\n",
    "    results_df = pd.DataFrame(test_results, columns=['Bin', 'P-Value', 'Significance'])\n",
    "    \n",
    "    return results_df\n",
    "\n",
    "# Perform the Mann Whitney U-test.\n",
    "mannwhitney_results = perform_mannwhitney_u_test(final_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to plot the results of the Mann-Whitney test\n",
    "def plot_mannwhitney_results(df, results_df, plot_output_path, category_colors='deep', top_n_bins=10):\n",
    "    # Filter for significant bins and select a subset for visualization\n",
    "    significant_bins = results_df[results_df['P-Value'] < 0.05].sort_values(by='P-Value', ascending=True)\n",
    "    significant_bins_subset = significant_bins.head(top_n_bins)\n",
    "    bins_order = significant_bins_subset['Bin'].tolist()\n",
    "    \n",
    "    # Filter original dataset for bins of interest\n",
    "    df_significant_subset = df[df['bin'].isin(bins_order)]\n",
    "    \n",
    "    # Plotting with ordered bins\n",
    "    plt.figure(figsize=(14, 8))\n",
    "    ax = sns.boxplot(x='bin', y='log10(methRatio)', hue='Category', \n",
    "                     data=df_significant_subset, palette=category_colors, order=bins_order)\n",
    "    plt.title(' ', fontsize=14)\n",
    "    plt.xticks(fontsize = 12, rotation=35)\n",
    "    ax.set_xlabel(' ', fontsize=14)\n",
    "    ax.set_ylabel('Methylation', fontsize=14)\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left', fontsize= 12)\n",
    "    plt.ylim(-1.5, 3.5)\n",
    "    plt.grid(True, axis='y')\n",
    "    \n",
    "    # Add grey blocks and annotations for significance\n",
    "    for i, bin in enumerate(bins_order):\n",
    "        if i % 2 == 0:\n",
    "            plt.axvspan(i - 0.5, i + 0.5, color='grey', alpha=0.2)\n",
    "        annotation = results_df.loc[results_df['Bin'] == bin, 'Significance'].values[0]\n",
    "        if annotation:\n",
    "            ax.text(i, 3, annotation, ha='center', va='bottom', color='black', fontsize=18)\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    sns.despine()\n",
    "    plt.savefig(plot_output_path, dpi=300)\n",
    "    plt.show()\n",
    "\n",
    "# Plot the results of the Mann Whitney U-test\n",
    "mannwhitney_boxplot_output_path = plots_directory/\"significant_bins_boxplot.jpeg\"  # Define your output path\n",
    "significant_bins = plot_mannwhitney_results(final_df, mannwhitney_results, mannwhitney_boxplot_output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to extract genomic location for a given bin\n",
    "def get_genomic_location(bin_name):\n",
    "    for (chrom, start, end), bin_id in bin_name_map.items():\n",
    "        if bin_id == bin_name:\n",
    "            return f\"{chrom}:{start}-{end}\"\n",
    "    return \"Location not found\"\n",
    "\n",
    "# Function to extract the bin name without the index number\n",
    "def extract_bin_name(full_bin):\n",
    "    return full_bin[full_bin.index('chr'):]\n",
    "\n",
    "# Create a copy of the significant results\n",
    "significant_results = mannwhitney_results[mannwhitney_results['P-Value'] < 0.05].copy()\n",
    "\n",
    "# Now apply the functions to create new columns\n",
    "significant_results.loc[:, 'Bin_Name'] = significant_results['Bin'].apply(extract_bin_name)\n",
    "significant_results.loc[:, 'Genomic_Location'] = significant_results['Bin_Name'].apply(get_genomic_location)\n",
    "\n",
    "# Reorder the columns for better readability\n",
    "significant_results = significant_results[['Bin', 'Bin_Name', 'Genomic_Location', 'P-Value', 'Significance']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to extract chromosome, start, and end from Genomic_Location\n",
    "def parse_genomic_location(location):\n",
    "    match = re.match(r'(chr\\w+):(\\d+)-(\\d+)', location)\n",
    "    if match:\n",
    "        return match.groups()\n",
    "    return None, None, None\n",
    "\n",
    "# Apply the parsing function and create new columns\n",
    "significant_results[['Chromosome', 'Start', 'End']] = significant_results['Genomic_Location'].apply(parse_genomic_location).apply(pd.Series)\n",
    "\n",
    "# Create BED file content\n",
    "bed_content = significant_results[['Chromosome', 'Start', 'End', 'Bin_Name', 'P-Value', 'Significance']]\n",
    "\n",
    "# Save to BED file\n",
    "bed_file_path = 'top_bins.bed'\n",
    "bed_content.to_csv(bed_file_path, sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Detect the genomic ranges of the top 10 significant bins and store the results in a pandas dataframe.\n",
    "if mannwhitney_results is not None:\n",
    "    mannwhitney_results.sort_values(by='P-Value', ascending=True, inplace=True)\n",
    "    top_10_significant_bins = mannwhitney_results.head(10)\n",
    "    top_10_significant_bins = bins_df[bins_df['bin_name'].isin(top_10_significant_bins['Bin'])]\n",
    "\n",
    "top_10_significant_bins_df = pd.DataFrame(top_10_significant_bins)\n",
    "top_10_significant_bins_df.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to the directory containing BED files\n",
    "reference_bed_path= '/Volumes/Intenso_SSD/Methylation_data/Reference_BED'\n",
    "\n",
    "# The BED file to intersect with all others\n",
    "bin_bed_file = pybedtools.BedTool('/Volumes/Intenso_SSD/Methylation_data/Reference_BED/top10_significant_bins_coordinates.bed')\n",
    "\n",
    "# Pattern to match all BED files in the directory\n",
    "bed_files_pattern = reference_bed_path + '/*.bed'\n",
    "\n",
    "# File to save the final results\n",
    "final_results_file = 'final_intersection_results.txt'\n",
    "\n",
    "# Initialize an empty list to hold the results\n",
    "all_results = []\n",
    "\n",
    "# List all BED files in the directory\n",
    "bed_files = glob.glob(bed_files_pattern)\n",
    "\n",
    "# Iterate over each BED file in the directory\n",
    "for bed_file in bed_files:\n",
    "    # Skip the output.bed file itself\n",
    "    if bed_file.endswith('top10_significant_bins_coordinates.bed'):\n",
    "        continue\n",
    "    \n",
    "    # Load the current BED file\n",
    "    reference_bed = pybedtools.BedTool(bed_file)\n",
    "    \n",
    "    # Perform the intersection\n",
    "    intersection_result = bin_bed_file.intersect(reference_bed, wb=True)\n",
    "    \n",
    "    # Extract and store the selected columns from the intersection\n",
    "    for feature in intersection_result:\n",
    "        fields = str(feature).strip().split('\\t')\n",
    "        if len(fields) >= 8:\n",
    "            all_results.append([fields[3], fields[7]])\n",
    "\n",
    "# Save the aggregated results to the final text file\n",
    "with open(final_results_file, 'w') as outfile:\n",
    "    for col in all_results:\n",
    "        outfile.write('\\t'.join(col) + '\\n')\n",
    "\n",
    "# Structure to hold the categorized RNAs by bin\n",
    "rna_bins = defaultdict(lambda: {'circRNA': [], 'miRNA': [], 'lncRNA': []})\n",
    "\n",
    "# Open the file and process each line\n",
    "with open(final_results_file, 'r') as file:\n",
    "    for line in file:\n",
    "        bin_id, rna_id = line.strip().split()\n",
    "        if rna_id.startswith('hsa_circ'):\n",
    "            rna_bins[bin_id]['circRNA'].append(rna_id)\n",
    "        elif rna_id.startswith('hsa-mir') or rna_id.startswith('hsa-miR'):\n",
    "            rna_bins[bin_id]['miRNA'].append(rna_id)\n",
    "        else:  # Assuming the remaining are lncRNAs based on provided data structure\n",
    "            rna_bins[bin_id]['lncRNA'].append(rna_id)\n",
    "\n",
    "\n",
    "# Prepare the final results for each bin\n",
    "final_results_data = {'Bin': [], 'circRNAs': [], 'miRNAs': [], 'lncRNAs': []}\n",
    "for bin_id, details in rna_bins.items():\n",
    "    final_results_data['Bin'].append(bin_id)\n",
    "    final_results_data['circRNAs'].append(', '.join(details['circRNA']))\n",
    "    final_results_data['miRNAs'].append(', '.join(details['miRNA']))\n",
    "    final_results_data['lncRNAs'].append(', '.join(details['lncRNA']))\n",
    "\n",
    "rna_interactions_df = pd.DataFrame(final_results_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set chromosomes\n",
    "circle = Gcircle(figsize=(14,14)) \n",
    "with open(\"/Volumes/Intenso_SSD/Methylation_data/general_chromosomes_human.csv\") as f:\n",
    "    f.readline()\n",
    "    for line in f:\n",
    "        line   = line.rstrip().split(\",\") \n",
    "        name   = line[0]\n",
    "        length = int(line[-1]) \n",
    "        arc    = Garc(arc_id=name, size=length, interspace=2, raxis_range=(735,785), labelposition=80, label_visible=True, labelsize=16)\n",
    "        circle.add_garc(arc) \n",
    "\n",
    "circle.set_garcs(0,360) \n",
    "for arc_id in circle.garc_dict:\n",
    "    circle.tickplot(arc_id, raxis_range=(785,800), tickinterval=20000000, ticklabels=None) \n",
    "\n",
    "#cytoband\n",
    "\n",
    "color_dict   = {\"gneg\":\"#FFFFFF00\", \"gpos25\":\"#EEEEEE\", \"gpos50\":\"#BBBBBB\", \"gpos75\":\"#777777\", \"gpos100\":\"#000000\", \"gvar\":\"#FFFFFF00\", \"stalk\":\"#C01E27\", \n",
    "               \"acen\":\"#D82322\"}\n",
    "\n",
    "arcdata_dict = defaultdict(dict)\n",
    "with open(\"/Volumes/Intenso_SSD/Methylation_data/karyotype_human.csv\") as f:\n",
    "    f.readline()\n",
    "    for line in f:\n",
    "        line  = line.rstrip().split(\",\")\n",
    "        name  = line[0]     \n",
    "        start = int(line[1])-1 \n",
    "        width = int(line[2])-(int(line[1])-1) \n",
    "        if name not in arcdata_dict:\n",
    "            arcdata_dict[name][\"positions\"] = []\n",
    "            arcdata_dict[name][\"widths\"]    = [] \n",
    "            arcdata_dict[name][\"colors\"]    = [] \n",
    "        arcdata_dict[name][\"positions\"].append(start) \n",
    "        arcdata_dict[name][\"widths\"].append(width)\n",
    "        arcdata_dict[name][\"colors\"].append(color_dict[line[-1]])\n",
    "\n",
    "for key in arcdata_dict:\n",
    "    circle.barplot(key, data=[1]*len(arcdata_dict[key][\"positions\"]), positions=arcdata_dict[key][\"positions\"], \n",
    "                   width=arcdata_dict[key][\"widths\"], raxis_range=[735,785], facecolor=arcdata_dict[key][\"colors\"]) \n",
    "\n",
    "df = methylation_df\n",
    "\n",
    "# Load seaborn deep palette\n",
    "palette = sns.color_palette(\"deep\")\n",
    "responders_color_pos = palette[0]  # Deep blue for positive values\n",
    "responders_color_neg = sns.light_palette(responders_color_pos, n_colors=3)[1]  # Lighter blue for negative values\n",
    "non_responders_color_pos = palette[1]  # Deep red for positive values\n",
    "non_responders_color_neg = sns.light_palette(non_responders_color_pos, n_colors=3)[1]  # Lighter red for negative values\n",
    "\n",
    "# Filter data for responders and non-responders\n",
    "responders = df[df['Category'] == 'Responders']\n",
    "non_responders = df[df['Category'] == 'Non_Responders']\n",
    "\n",
    "# Function to add bar plots for positive and negative values separately\n",
    "def add_barplots(data, chrom, radial_range_pos, radial_range_neg, color_pos, color_neg):\n",
    "    positions = [(start + end) // 2 for start, end, _ in data]\n",
    "    meth_ratios = [log10_meth_ratio for _, _, log10_meth_ratio in data]\n",
    "    widths = [end - start for start, end, _ in data]\n",
    "    \n",
    "    # Separate positive and negative log10(methRatio) values\n",
    "    pos_data = [val if val >= 0 else 0 for val in meth_ratios]\n",
    "    neg_data = [-val if val < 0 else 0 for val in meth_ratios]\n",
    "    \n",
    "    # Plot positive values (hypermethylated)\n",
    "    circle.barplot(chrom, data=pos_data, positions=positions, width=widths, raxis_range=radial_range_pos, facecolor=color_pos, edgecolor=color_pos, linewidth=0.05)\n",
    "    \n",
    "    # Plot negative values (hypomethylated)\n",
    "    circle.barplot(chrom, data=neg_data, positions=positions, width=widths, raxis_range=radial_range_neg, facecolor=color_neg, edgecolor=color_neg, linewidth=0.05)\n",
    "\n",
    "# Prepare data for responders\n",
    "responder_data = defaultdict(list)\n",
    "for _, row in responders.iterrows():\n",
    "    chrom = row['chrom']\n",
    "    start = row['chromStart']\n",
    "    end = row['chromEnd']\n",
    "    log10_meth_ratio = row['log10(methRatio)']\n",
    "    responder_data[chrom].append((start, end, log10_meth_ratio))\n",
    "\n",
    "# Prepare data for non-responders\n",
    "non_responder_data = defaultdict(list)\n",
    "for _, row in non_responders.iterrows():\n",
    "    chrom = row['chrom']\n",
    "    start = row['chromStart']\n",
    "    end = row['chromEnd']\n",
    "    log10_meth_ratio = row['log10(methRatio)']\n",
    "    non_responder_data[chrom].append((start, end, log10_meth_ratio))\n",
    "\n",
    "# Add bar plots for responders\n",
    "for chrom, values in responder_data.items():\n",
    "    add_barplots(values, chrom, radial_range_pos=(650, 730), radial_range_neg=(580, 640), color_pos='blue', color_neg=responders_color_neg)\n",
    "\n",
    "# Add bar plots for non-responders\n",
    "for chrom, values in non_responder_data.items():\n",
    "    add_barplots(values, chrom, radial_range_pos=(510, 570), radial_range_neg=(440, 500), color_pos='orange', color_neg=non_responders_color_neg)   \n",
    "\n",
    "# Add label to the center of the circos plot\n",
    "plt.text(0.5, 0.5, 'Genome Wide Methylation Profile of \\n Responders (blue) and Non Responders (orange)', horizontalalignment='center', verticalalignment='center', \n",
    "         transform=circle.figure.transFigure, fontsize=15)\n",
    "\n",
    "# Show the final plot\n",
    "circle.figure.savefig('circos.png', dpi=300)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
